### PR TITLE
Add d3-sankey Definitions

### LINKED ISSUE #14797: Definitely typed file for d3 sankey plugin
Can someone publish a definitelytyped file for d3-sankey plugin?
Here is the link for the plugin
https://github.com/d3/d3-sankey/

I am relatively new to typescript. Otherwise, would have done that myself.

Thanks in advance!

### PR DESCRIPTION
**Pull request by tomwanzek**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

This PR adds definitions for **d3-sankey** the folder structure corresponds to the one used for the core D3 module definitions.

In reviewing the definition, please be advised, that it is based on a source code review of [d3-sankey](https://github.com/d3/d3-sankey/blob/master/src/sankey.js). This approach was chosen, as the current API documentation of the repo misses key methods of the methods of the Sankey Layout generator. The documentation also makes misguiding statements about the use of accessor functions, when none are used in the code.

I will open a separate issue in the d3-sankey repo to ask for changes to be made and submit a PR there, if agreeable.

Closes #14797.

Kindly asking for a review @gustavderdrache @Ledragon 
cc @ravibha

### CODE REVIEW

**Comment by tomwanzek** [comment]
FYI, I opened [d3-sankey Issue 22](https://github.com/d3/d3-sankey/issues/22) to raise the API Doc/source disconnect.

**Comment by tomwanzek** [comment]
@gustavderdrache, @Ledragon if none of you currently has the cycles to review the PR with your always appreciated second set of eyes, I would make the following  proposal in the interest of getting this out the door:
* Just quickly ping back to indicate availability or lack thereof by Sunday this week,
* If otherwise, no-one has cycles or objects by then, I would suggest to proceed with the merge.

I have tested out the definitions in an actual Angular 2 project by manually bringing them into the compilation scope and they work without issue. 

In the meantime I also submitted tan API Doc update PR to **d3-sankey** itself (see [here](https://github.com/d3/d3-sankey/pull/23))

Co-maintainer additions for the definitions in the header are available as always, if you're interested. 😉

File: types/d3-sankey/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+// Last module patch version validated against: 0.4.2
+
+/**
+ * A helper interface as an extension reference for user-provided properties of
+ * nodes and links in the graph, which are not required or calculated by
+ * the Sankey Layout Generator
+ */
+export interface SankeyExtraProperties { [key: string]: any; }
```
Comment (gustavderdrache) [inline types/d3-sankey/index.d.ts]: This interface seems dangerous. There aren't any default properties, so why not just specify the type parameters in e.g., `SankeyNodeMinimal` as just `<N, L>`?

FWIW, I deleted the interface and none of the tests fail.

File: types/d3-sankey/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+export type SankeyLink<N extends SankeyExtraProperties, L extends SankeyExtraProperties> = L & SankeyLinkMinimal<N, L>;
+
+/**
+ * An svg path generator factory for the link paths in a calculated Sankey Layout.
+ */
+export interface SankeyLinkPathGenerator<N extends SankeyExtraProperties, L extends SankeyExtraProperties> {
+    /**
+     * Return svg path string for a given link.
+     *
+     * IMPORTANT: Only invoke for link data with Sankey Layout information previosuly calculated.
```
Comment (gustavderdrache) [inline types/d3-sankey/index.d.ts]: Typo: "previosuly" should be "previously".

**Comment by gustavderdrache** [comment]
Apologies for the delay, I was on vacation recently and am still catching up on my backlog a bit.

**Comment by tomwanzek** [comment]
No worries and good for you 🕺 . That's why I said, a simple ping with do or don't have cycles is fine. The module is not a key one in the ecology and still in pre-1.0.0 version. It's also not as frequently loaded according to npmjs. That's why I would not be to worried merging it as is.

The most interesting thing was the design challenge around the node and link interfaces. They are even more interdependent than the **d3-force** ones. So in turn the mental model for the generics is somewhat different as well.

File: types/d3-sankey/index.d.ts (hunk 2 above)
Comment (tomwanzek) [inline types/d3-sankey/index.d.ts]: Will fix. Cheers.

File: types/d3-sankey/index.d.ts (hunk 1 above)
Comment (tomwanzek) [inline types/d3-sankey/index.d.ts]: Very good question. I wasn't to happy with it either. Here is the rationale:

It prevents silly things like:  `SankeyNode<number, string>`, i.e. using primitives instead of something that has a key-value structure.

So you're right, dropping it won't fail any tests or meaningful use, it only prevents an edge case failure mode.

(_Side note_: Sooner or later we might want to consider opening up a discussion issue regarding creating and publishing a TS2.2+ conforming definition, e.g. here we could use default assignments of `{}` for the generics. I just dread 😰  the parallel maintenance period of the definitions, even if we can shorten some signatures, consider using `keyof` etc.)

Comment (gustavderdrache) [inline types/d3-sankey/index.d.ts]: Fair enough. I've played with it and it doesn't seem like it introduces any accidental issues, so I'm down.

**Comment by gustavderdrache** [comment]
:+1:

**Comment by Ledragon** [comment]
I'm afraid I won't be able to invest the time to provide proper constructive feedback... If I manage to review, I will, but don't wait for me to merge

**Comment by tomwanzek** [comment]
Spoils to the winner of the speediest review competition 😉 .

**Comment by tomwanzek** [comment]
@sheetalkamat Could you be so kind an trigger the merge. Thx, T.
